[JENKINS-36084] Responsive new job name validation

samatdav opened the pull request:
Currently the GET request to check new job name for invalid characters is sent only on 'blur' jQuery event.
That makes a user know about the invalid character only after he entered the name and forces to go back to input field.

This change checks for invalid characters in new job name after each character (keyup event).
Additionally, it removes unnecessary warning messages when job name is empty.

samatdav committed:
Improved new job name validation

samatdav committed:
delete space

samatdav committed:
removed focus and unneccesary client side check

samatdav committed:
new line at the end

samatdav committed:
sidepanel popup change

samatdav committed:
comment fix

samatdav committed:
dirty popup with library

samatdav committed:
Removed unnecessary st:include tags

samatdav committed:
remove st:include

samatdav committed:
Added newJob properties to sidepanel properties to fix labels

samatdav committed:
uncomment popup

samatdav committed:
romoved Bootstrap from newJob

samatdav committed:
romoved unnecessary popup jelly

samatdav committed:
romoved unnecessary popup jelly

samatdav committed:
changed popup.css

samatdav committed:
put remodal popup styles to remodal.css

samatdav committed:
added remodal.css

samatdav committed:
moved my rule to addform.less

samatdav committed:
moved popup from sidepanel.jelly to popup.jelly

samatdav committed:
changed the look of new job creation

samatdav committed:
Revert "Improved new job name validation"

This reverts commit ed80c35d3bcc1dd1230a48116b293f4bc94ae3b9.

samatdav committed:
Revert "delete space"

This reverts commit d2dff11357886105edf141940a49607dacd6018a.

samatdav committed:
commiting changes

samatdav committed:
Revert "Revert "Improved new job name validation""

This reverts commit 4a3f75ac675499f8058956b2ad2aa1217d10e78d.

samatdav committed:
commit files

samatdav committed:
removed UI changes

samatdav committed:
moved properties and removed name validation change

samatdav committed:
fixed mistakes

samatdav committed:
removed task creation on new page

samatdav committed:
check on keyup

samatdav committed:
removed popup

samatdav committed:
removed popup fix

samatdav committed:
different handeling of keyup and blur

michaelneale commented:
looks like jshint error now: 

`03:06:32 [ERROR] 
03:06:32 [ERROR] events.js:141
03:06:32 [ERROR]       throw er; // Unhandled 'error' event
03:06:32 [ERROR]       ^
03:06:32 [ERROR] Error: JSHint failed for: src/main/js/add-item.js
`

samatdav commented:
@michaelneale can you tell how to approach this JSHint error?
Also I don't understand why the remote check fails, but shows that all tests passed: https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/5484/testReport/

samatdav committed:
Merge remote-tracking branch 'upstream/master' into new-job-name-validation

samatdav committed:
fix JSHint errors

oleg-nenashev commented:
Is it WiP?

samatdav commented:
@oleg-nenashev I fixed the issued I wanted to solve and now do not see problems. So now I plan to ask mentors (and may be others) if they have other comments and if the change looks good to them.

samatdav commented:
@lanwen @recena do you have comments on the change? What it needs to be approved?

recena commented:
@samatdav I'd like to review this PR.

recena commented:
@samatdav If you are working on [JENKINS-36084](https://issues.jenkins-ci.org/browse/JENKINS-36084) it is helpful if you keep the JIRA issue status updated.

oleg-nenashev commented:
Modified the PR status according to the feedback

lanwen commented on war/src/main/js/add-item.js:
```diff
@@ -246,15 +245,23 @@ $.when(getItems()).done(function(data) {
             setFieldValidationStatus('name', true);
             if (getFormValidationStatus()) {
-              enableSubmit(true);
+              if (event.type === 'blur') { 
```
maybe it will be look better with switch-case

lanwen commented on war/src/main/js/add-item.js:
```diff
@@ -246,15 +245,23 @@ $.when(getItems()).done(function(data) {
+              if (event.type === 'blur') { 
+                enableSubmit(true);
+              } else if (event.type === 'keyup') {
+                var btn = $('form .footer .btn-decorator button[type=submit]');
```
what does it mean?

samatdav replied:
Since enable submit contains btn.focus() in it, that is to disable the focus of the button for keyup event. So when event was keyup, only necessary part of enable submit is called.

samatdav replied:
The btn variable defined again since it was defined only locally before and needed again now.

lanwen replied:
can u reuse some code to define this action? It looks now like a temp solution

daniel-beck replied:
Would it be possible to separate enabling and focusing, and re-use enabling here?

samatdav committed:
cleanup

samatdav committed:
added single quote

oleg-nenashev commented:
ping @samatdav

samatdav committed:
separated button enabling and focus

samatdav committed:
cleanup

samatdav commented on war/src/main/js/add-item.js:
```diff
@@ -246,15 +257,22 @@ $.when(getItems()).done(function(data) {
+                  enableSubmit(true);
+                  break;
+                case 'keyup':
+                  enableButton(true);
```
@daniel-beck @lanwen Is this how you suggested to separate focusing part of enableSubmit feature?

daniel-beck replied:
@samatdav Pretty much, yes.

batmat commented:
Also ping @jenkinsci/code-reviewers

daniel-beck commented:
One weird interaction. This stupid form still has autocompletion support, so any time I enter a job name similar to an existing job name, I get the stupid dropdown from my browser that hides the validation area.

Previously, this wasn't a big deal, as validation only kicked in after leaving the input field. Now, validation is shown immediately, but hidden by the dropdown.

This could (maybe even should) be addressed in the context of this PR to make this change actually useful -- as I expect name duplication to be more common than using terrible characters, and there's an overlap here.

What do others think?

michaelneale commented:
@daniel-beck as in it pops up suggestions for new names based on existing names? that seems really out of place here. The only time that could be useful would be if you are establishing a naming convention and it makes it easy to stick to things, otherwise shouldn't that be turned off?

daniel-beck commented:
@michaelneale 

> otherwise shouldn't that be turned off? 

It's a web browser feature. This is one of the few form fields in Jenkins that don't have this disabled, and the problem is that the dropdown overlaps the validation error.

I.e. while this PR appears to accomplish everything it set out to do, it's not working great due to the browser's form field autocompletion feature that gets into the way

michaelneale commented:
@samatdav it seems that `autocomplete=off` needs to be an attribute set somewhere as per https://developer.mozilla.org/en-US/docs/Web/Security/Securing_your_site/Turning_off_form_autocompletion for that name field (I don't think people would object)

oh dear: 

> In some case, the browser will keep suggesting autocompletion values even if the autocomplete attribute is set to off. This unexpected behavior can be quite puzzling for developers. The trick to really force the no-completion is to assign a random string to the attribute like so :
> 
> autocomplete="nope"
>

samatdav commented:
@tfennelly @recena Could you provide feedback on both the proposed change and the autocomplete issue (in last comments)? Do you suggest to change anything?

tfennelly commented on war/src/main/js/add-item.js:
```diff
@@ -116,6 +116,18 @@ $.when(getItems()).done(function(data) {
+      var btn = $('form .footer .btn-decorator button[type=submit]');
+      if (status === true) {
+        btn.focus();
+      }
```
Should there be an `else` here to remove focus?

tfennelly replied:
Also ... as things are, there's no need to do the jquery select until inside the `if` block i.e. at the moment it's redundant what `status === false`.

samatdav replied:
`else` functionality was not present initially and seems that there may not be much need for it. Or do you think it would be better with `else` ?

Ok, I moved selecting inside if block.

tfennelly commented on war/src/main/js/add-item.js:
```diff
@@ -116,6 +116,18 @@ $.when(getItems()).done(function(data) {
     }
 
+    function focusButton(status) {
+      var btn = $('form .footer .btn-decorator button[type=submit]');
```
I think add a `getFormSubmitButton` function that hides the jquery selector and can be reused here and in the `enableButton` function.

samatdav replied:
I got an error trying to use 
```
function getFormSubmitButton() {
    var btn = $('form .footer .btn-decorator button[type=submit]');
}
```
and then just calling it inside.
Do you think it is necessary to replace it?

tfennelly replied:
Try ...

```javascript
function getFormSubmitButton() {
    return $('form .footer .btn-decorator button[type=submit]');
}
```

And then to use it...

```javascript
var btn = getFormSubmitButton();

// etc...
```

tfennelly commented on war/src/main/js/add-item.js:
```diff
@@ -100,14 +100,14 @@ $.when(getItems()).done(function(data) {
-        btn.focus();
       } else {
         if (!btn.hasClass('disabled')) {
           btn.addClass('disabled');
```
I know the above code is not part of this PR, but it could be cleaned up. The `btn.hasClass('disabled')` calls are dumb. I'm sure the above code could be more simply stated as ...

```javascript
function enableButton(status) {
  var btn = $('form .footer .btn-decorator button[type=submit]');

  btn.prop('disabled', !status);
  if (status === true) {
    btn.removeClass('disabled');
  } else {
    btn.addClass('disabled');
  }
}
```

samatdav replied:
Thanks, I added your code.

samatdav committed:
Tom's suggestions

samatdav commented:
@tfennelly I implemented most of your suggestions. Do you have any other objections? Do you have an opinion about autocomplete?

tfennelly commented:
@samatdav wrt autocomplete ... assuming there'd be objections to removing it, how about moving the validation warning so as the autocomplete doesn't hide it?

samatdav commented:
@tfennelly I am afraid that will create even more debate on whether and how the validation warning should be moved.
I plan to make a PR with UI changes but I am afraid it will take forever to discuss. 
Here are some of my UI ideas. May be a part of it would be considered viable.
https://github.com/samatdav/jenkins/pull/2/files

samatdav commented:
@tfennelly In https://github.com/jenkinsci/jenkins/pull/2501 I propose to move input box so that autocomplete will not hide warning message. What do you think?

samatdav committed:
moved invalid warning

lanwen commented:
How it looks like now?

samatdav commented:
@lanwen 
![out-3](https://cloud.githubusercontent.com/assets/9495022/17617701/92c05ffa-608c-11e6-8252-b1d800930090.gif)

daniel-beck commented:
@tfennelly WDYT? Can't help it, this looks weird…

samatdav committed:
red border

samatdav commented:
@daniel-beck @lanwen @michaelneale I removed the shift and added a red border. What do you think?
![out-4](https://cloud.githubusercontent.com/assets/9495022/17626533/a96e7e80-60be-11e6-9fa6-b27b73176bf6.gif)

lanwen commented:
red border looks much better

daniel-beck commented:
The suggestion to indicate an invalid title through styling of the text field came up in today's meeting about Samat's GSoC project. I like it as well, but wonder whether a background color could make it even more noticeable without distracting so much (you want people to notice, after all!)

That said, even if it's just the border, I'm now 👍 on this. Further refinements can be done later, and Samat got the live validation done.

daniel-beck commented on war/src/main/js/add-item.js:
```diff
@@ -234,27 +240,36 @@ $.when(getItems()).done(function(data) {
+            $('#add-item-panel input').css('border', "1px solid red");
           } else {
             cleanValidationMessages('.add-item-name');
+            $('#add-item-panel input').css('border', "1px solid #999");
```
Can't this use CSS classes that already exist?

samatdav replied:
@daniel-beck Is it better know? 
https://github.com/jenkinsci/jenkins/pull/2416/files#diff-e146d4fdbfe6385a456aea9775f6282dR250

samatdav replied:
![red](https://cloud.githubusercontent.com/assets/9495022/17659323/c582e7fc-62e2-11e6-9f1b-a3b20e298cfb.jpg)

samatdav committed:
red background

lanwen commented on war/src/main/js/add-item.js:
```diff
@@ -234,27 +240,36 @@ $.when(getItems()).done(function(data) {
           var message = parseResponseFromCheckJobName(data);
           if (message !== '') {
             activateValidationMessage('#itemname-invalid', '.add-item-name', message);
+            $('input[name="name"]').css({'border': "1px solid red", "background-color": "#FADEDE"});
```
don't add exact css, toggle class

samatdav committed:
class toggle

samatdav commented:
@lanwen I added class toggle. Is it fine now?

lanwen commented on war/src/main/js/widgets/add/addform.less:
```diff
@@ -14,6 +14,11 @@
     z-index: 5;
   }
 
+  .warning-input {
```
i'd suggest `job-name_incorrect`

samatdav committed:
renamed class

samatdav commented:
@lanwen I renamed class. Is it fine now?

samatdav commented:
@michaelneale can you restart the build? It seems like a remote failure.

lanwen commented:
LGTM

daniel-beck commented:
When clearing the input field, it retains the previous validation status. Is that deliberate?

samatdav commented:
@daniel-beck what do you mean by retains? If a recall correctly the behavior is the same as with blur event.

michaelneale commented:
@samatdav he means that if you delete the contents, it still shows an error around it (which is how I interpret what he said, but I don't think my explanation is any clearer).

samatdav commented:
@daniel-beck @michaelneale oh, I see - do you mean the red border still shows up after you delete the job name which is invalid? I am actually not sure what happens because the code for replacing warning message and removing warning class name is in one place: https://github.com/jenkinsci/jenkins/pull/2416/files#diff-e146d4fdbfe6385a456aea9775f6282dR253
While the warning text message is removed, the red border is not. May be you could suggest what may be the error and a possibile solution?

oleg-nenashev commented:
@michaelneale @samatdav I'd guess there is no interest to proceed with this change, right?
If no, I propose to close it.

daniel-beck commented:
Based on latest review comments, there's also an unresolved issue when clearing the input field.

Closing for now, can always be reopened or resubmitted later.
